lib/indexer/change_notification_processor.rb

```diff
@@ -0,0 +1,54 @@
+# ChangeNotificationProcessor
+#
+# This is called by the message queue processor and is responsible
+# for acting on the incoming payload. It checks if the updated document in the
+# payload needs to be updated, looks up the document in elasticsearch and triggers
+# a "re-index", which means that the document will go through the DocumentPreparer
+# process and will look up the tags for the document again.
+module Indexer
+  class ChangeNotificationProcessor
+    EXCLUDED_FORMATS = %w{email_alert_signup gone redirect}.freeze
+
+    # Gets initialised with a content-item from the publishing-api message queue.
+    def initialize(content_item)
+      @content_item = content_item
+    end
+
+    def trigger
+      return if should_skip_document?
+      trigger_indexing_of_document
+    rescue KeyError => e
+      raise Indexer::MalformedMessage, "Content item attribute missing. #{e.message}"
+    end
+
+  private
+
+    attr_reader :content_item
+
+    def document
+      @document ||= begin
+        document_base_path = content_item.fetch("base_path")
+        unified_index = search_server.index_for_search(search_config.content_index_names)
+        document = unified_index.get_document_by_link(document_base_path)
+        document || raise(Indexer::UnknownDocumentError, "Document not found in index")
+      end
+    end
+
+    def trigger_indexing_of_document
+      index = search_server.index(document['real_index_name'])
+      index.trigger_document_reindex(document['_id'])
+    end
+
+    def should_skip_document?
+      EXCLUDED_FORMATS.include?(content_item.fetch("document_type"))
+    end
+
+    def search_server
+      @search_server ||= search_config.search_server
+    end
+
+    def search_config
+      @search_config ||= Rummager.settings.search_config
```

Comment:
Yeah, the way this indexes things fits together is horrible. I've extracted the index selecting into a little class to hide the ugliness: https://github.com/alphagov/rummager/pull/632/commits/d71c51aaf171ef58c528bfa54073f89cf04425b6

This is not a good solution, but at least it moves the problem away a bit. A real solution would involve refactoring the config/index in tandem with the elasticsearch-client stuff, like @MatMoore says.